### PR TITLE
✨Validates CoreDNS version update

### LINKED ISSUE #2620: Add coredns validation in KCP validation webhook
**User Story**

As an operator, I would like to ensure my coredns version update is "migratable" by having extra validation in the webhook.

**Detailed Description**

In KCP KubeadmConfig, there is a property
```
dns:
...
  ImageTag: 1.2.3-foobar.1
```
We need to have validation in the webhook that takes the `ImageTag` and make sure it is valid using the `github.com/coredns/migration` library. 
I believe there is a `migration.validUpMigration` method that does it but unfortunately it is not exported. 

**Anything else you would like to add:**

Currently coredns upgrade code calls the `migration.Migrate` which does migrate up and not migrate down. So that means we are only handling upwards migrations at the moment.

/kind feature
/area control-plane

### PR DESCRIPTION
**Pull request by gab-satchi**
- Skips validation if ClusterConfiguration undefined
- Skips validation if previous version undefined

**What this PR does / why we need it**:
This PR validates CoreDNS upgrade compatibility in the KCP webhook. It utilizes a function in the coredns-migration library that will compare two versions and determine if it can be upgraded from one to the other. 

**Which issue(s) this PR fixes** 
Fixes #2620

### CODE REVIEW

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 1]
```diff
@@ -259,6 +262,33 @@ func (in *KubeadmControlPlane) validateCoreDNS() (allErrs field.ErrorList) {
...
+		return allErrs
+	}
+	dns := &in.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+	//return if the type is anything other than empty (default), or CoreDNS.
+	if dns.Type != "" && dns.Type != kubeadmv1.CoreDNS {
+		return allErrs
+	}
+
+	fromVersion := prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag
+	if err := migration.ValidUpMigration(fromVersion, dns.ImageTag); err != nil {
```
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go]: See #2697, there is a different issue where we want to allow upgrading for example from v1.0.6-foobar.1 to v1.0.6-foobar.2, there is some good context in the PR's comments and how we approached this in code.

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 2]
```diff
@@ -259,6 +262,55 @@ func (in *KubeadmControlPlane) validateCoreDNS() (allErrs field.ErrorList) {
...
+			field.Invalid(
+				field.NewPath("spec", "kubeadmConfigSpec", "clusterConfiguration", "dns", "imageTag"),
+				dns.ImageTag,
+				fmt.Sprintf("failed to parse CoreDNS target version: %v", dns.ImageTag),
+			),
+		)
+		return allErrs
+	}
+
+	if err := migration.ValidUpMigration(fromVersion.String(), toVersion.String()); err != nil {
```
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go]: Can we allow for the version major.minor.patch to be the same but the actual strings are different?

For example: `v0.1.6-foo.1` and `v0.1.6-foo.2` have the same version, but does the `ValidUpMigration` return an error in that case?

Also let's add a unit test that covers this case

Comment (gab-satchi) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go]: ValidUpMigration doesn't return an error in that case. https://github.com/kubernetes-sigs/cluster-api/pull/2707/files#diff-0d49cc99aec2899fcc03b149561ff8eeR305-R311 this test should cover that case.  It tries to update from `v1.6.6_foobar.1` to `v1.6.6_foobar.2`

File: go.mod [hunk 3]
```diff
@@ -5,7 +5,7 @@ go 1.13
 require (
 	github.com/MakeNowJust/heredoc v1.0.0
 	github.com/blang/semver v3.5.1+incompatible
-	github.com/coredns/corefile-migration v1.0.6
+	github.com/coredns/corefile-migration v1.0.7-0.20200317205912-4b5f6e751512
```
Comment (vincepri) [inline go.mod]: We should probably stick with the latest released version, is there any reason this was updated to a commit?

Comment (vincepri) [reply go.mod]: @rajansandeep are you folk planning to release 1.0.7 soon-ish?

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go (hunk 2 above)
Comment (vincepri) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go]: I see it's covered here https://github.com/coredns/corefile-migration/blob/master/migration/migrate.go#L424, great work :)

File: go.mod (hunk 3 above)
Comment (gab-satchi) [reply go.mod]: Yeah I was wondering the same. I updated it to the commit that exported `ValidUpMigration` with the intent for it to be temporary.

Comment (rajansandeep) [reply go.mod]: @vincepri We usually have a release every upstream k8s cycle, but if it's necessary we can make a release for this case

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go (hunk 2 above)
Comment (rajansandeep) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go]: For the case of the coredns migration, if the from to to version is the same, we just skip migrating the corefile:

https://github.com/coredns/corefile-migration/blob/master/migration/migrate.go#L129-L131

File: go.mod (hunk 3 above)
Comment (vincepri) [reply go.mod]: It'd be great to have a tag, if possible, so we don't have to use the hash commit here

Comment (rajansandeep) [reply go.mod]: Release 1.0.7 is now available 
https://github.com/coredns/corefile-migration/releases/tag/v1.0.7

Comment (vincepri) [reply go.mod]: 🎉 thank you!! @gab-satchi over to you!

Comment (gab-satchi) [reply go.mod]: Appreciate it @rajansandeep. Just updated @vincepri

**Pull request merged**